### PR TITLE
Car 2670: add cart to cc

### PR DESCRIPTION
**Pull request by ahalberkamp**
In this PR cart-data is added to the creditcard pre- and authorization request to improve the customer experience.
This was recommended by payone, please check: https://docs.payone.com/display/public/PLATFORM/Special+remarks+-+3-D+Secure

Jira: https://jira.commercetools.com/browse/CAR-2670

### CODE REVIEW

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java [hunk 1]
```diff
@@ -212,7 +218,7 @@ private PricesAggregator populatePrices(List<Supplier<PricesAggregator>> pricesV
     }
 
     /**
-     * Populate "goods" of the Klarna request. This function is used for {@link LineItem} of CTP.
+     * Populate "goods" of the  request. This function is used for {@link LineItem} of CTP.
```
Comment (heshamMassoud) [inline service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java]:
```suggestion
     * Populate "goods" of the request. This function is used for {@link LineItem} of CTP.
```

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java [hunk 2]
```diff
@@ -35,16 +37,17 @@
 import static java.util.Optional.ofNullable;
 
 /**
- * Base class for Klarna <i>authorisation</i> and <i>preauthorisation</i> requests.
+ * Base class for creditcard and Klarna <i>authorisation</i> and <i>preauthorisation</i> requests.
```
Comment (heshamMassoud) [inline service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java]:
```suggestion
 * Base class for credit card and Klarna <i>authorisation</i> and <i>preauthorisation</i> requests.
```

File: service/src/main/java/com/commercetools/pspadapter/payone/mapping/CreditCardRequestFactory.java [hunk 3]
```diff
@@ -50,7 +50,7 @@ public CreditCardAuthorizationRequest createAuthorizationRequest(
         Preconditions.checkArgument(ctPayment.getCustom() != null, "Missing custom fields on payment!");
 
         final String pseudocardpan = ctPayment.getCustom().getFieldAsString(CustomFieldKeys.CARD_DATA_PLACEHOLDER_FIELD);
-        CreditCardAuthorizationRequest request = new CreditCardAuthorizationRequest(getPayoneConfig(), pseudocardpan);
+        CreditCardAuthorizationRequest request = new CreditCardAuthorizationRequest(getPayoneConfig(), pseudocardpan,paymentWithCartLike);
```
Comment (heshamMassoud) [inline service/src/main/java/com/commercetools/pspadapter/payone/mapping/CreditCardRequestFactory.java]:
```suggestion
        CreditCardAuthorizationRequest request = new CreditCardAuthorizationRequest(getPayoneConfig(), pseudocardpan, paymentWithCartLike);
```

File: service/src/main/java/com/commercetools/pspadapter/payone/mapping/CreditCardRequestFactory.java [hunk 4]
```diff
@@ -33,7 +33,7 @@ public CreditCardPreauthorizationRequest createPreauthorizationRequest(
         Preconditions.checkArgument(ctPayment.getCustom() != null, "Missing custom fields on payment!");
 
         final String pseudocardpan = ctPayment.getCustom().getFieldAsString(CustomFieldKeys.CARD_DATA_PLACEHOLDER_FIELD);
-        CreditCardPreauthorizationRequest request = new CreditCardPreauthorizationRequest(getPayoneConfig(), pseudocardpan);
+        CreditCardPreauthorizationRequest request = new CreditCardPreauthorizationRequest(getPayoneConfig(), pseudocardpan,paymentWithCartLike);
```
Comment (heshamMassoud) [inline service/src/main/java/com/commercetools/pspadapter/payone/mapping/CreditCardRequestFactory.java]:
```suggestion
        CreditCardPreauthorizationRequest request = new CreditCardPreauthorizationRequest(getPayoneConfig(), pseudocardpan, paymentWithCartLike);
```

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/klarna/KlarnaPreauthorizationRequest.java [hunk 5]
```diff
@@ -2,12 +2,14 @@
...
+import com.commercetools.pspadapter.payone.domain.payone.model.common.ExtendedAuthorizationRequest;
 import com.commercetools.pspadapter.payone.domain.payone.model.common.RequestType;
 
-public class KlarnaPreauthorizationRequest extends BaseKlarnaRequest {
+public class KlarnaPreauthorizationRequest extends ExtendedAuthorizationRequest {
 
     public KlarnaPreauthorizationRequest(final PayoneConfig config, final String financingtype,
                                          final PaymentWithCartLike paymentWithCartLike) {
-        super(config, RequestType.PREAUTHORIZATION.getType(), financingtype, paymentWithCartLike);
+        super(config, RequestType.PREAUTHORIZATION.getType(), financingtype, ClearingType.PAYONE_KLV.getPayoneCode(),paymentWithCartLike);
```
Comment (heshamMassoud) [inline service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/klarna/KlarnaPreauthorizationRequest.java]:
```suggestion
        super(config, RequestType.PREAUTHORIZATION.getType(), financingtype, ClearingType.PAYONE_KLV.getPayoneCode(), paymentWithCartLike);
```

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java [hunk 6]
```diff
@@ -60,13 +63,16 @@
...
     private List<Integer> va;
 
-    protected BaseKlarnaRequest(final PayoneConfig config, final String requestType,
-                                final String financingtype, final PaymentWithCartLike paymentWithCartLike) {
-        super(config, requestType, ClearingType.PAYONE_KLV.getPayoneCode());
-
-        this.financingtype = financingtype;
-
-        mapKlarnaPropertiesFromPaymentCart(paymentWithCartLike);
+    protected ExtendedAuthorizationRequest(final PayoneConfig config,
```
Comment (heshamMassoud) [inline service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java]: Maybe it makes sense to add @nullable and @Nonnull annotations to clearly distinguish between params which could be null and those that can't.

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java [hunk 7]
```diff
@@ -60,13 +63,16 @@
...
+                                           final String requestType,
+                                           final String financingtype,
+                                           final String clearingType,
+                                           final PaymentWithCartLike paymentWithCartLike) {
+        super(config, requestType, clearingType);
+        if (StringUtils.isNotBlank(financingtype)) {
```
Comment (heshamMassoud) [inline service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java]: I think there is no need to check if the string is blank (`""`) here

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java [hunk 8]
```diff
@@ -35,16 +37,17 @@
...
- * {@link #mapKlarnaPropertiesFromPaymentCart(PaymentWithCartLike)} from the {@link PaymentWithCartLike}.
+ * {@link #mapPropertiesFromPaymentCart(PaymentWithCartLike)} from the {@link PaymentWithCartLike}.
  * <p>
- * See <i>PAYONE_Platform_Klarna_Addon_EN</i> for more details.
+ * See <i>https://docs.payone.com/display/public/PLATFORM/Special+remarks+-+3-D+Secure</i> and
+ * <i>PAYONE_Platform_Klarna_Addon_EN</i>   for more details.
  * The latest implemented API version is from <b>2016-11-30</b>.
  */
-public abstract class BaseKlarnaRequest extends AuthorizationRequest {
+public abstract class ExtendedAuthorizationRequest extends AuthorizationRequest {
```
Comment (heshamMassoud) [inline service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java]: Maybe a more self-descriptive name: `AuthorizationRequestWithCart`

File: service/src/test/java/com/commercetools/pspadapter/payone/mapping/MappingUtilTest.java [hunk 9]
```diff
@@ -57,11 +59,13 @@
...
 
     @Mock
-    private CartLike cardLike;
+    private CartLike cartLike;
 
     @Before
     public void setUp() throws Exception {
         super.setUp();
+        final PaymentTestHelper payments = new PaymentTestHelper();
+       when((CartLike) paymentWithCartLike.getCartLike()).thenReturn(payments.dummyOrderMapToPayoneRequest());
```
Comment (heshamMassoud) [inline service/src/test/java/com/commercetools/pspadapter/payone/mapping/MappingUtilTest.java]: indentation

File: service/src/test/java/com/commercetools/pspadapter/payone/mapping/CreditCardRequestFactoryTest.java [hunk 10]
```diff
@@ -185,6 +186,15 @@ public void createFullAuthorizationRequestFromValidPayment() throws Exception {
         softly.assertThat(result.getShipping_country()).isEqualTo(shippingAddress.getCountry().toLocale().getCountry());
         softly.assertThat(result.getShipping_state()).isEqualTo(shippingAddress.getState());
         softly.assertThat(result.getShipping_company()).isEqualTo(shippingAddress.getCompany() + " " + shippingAddress.getDepartment());
+        // but then below the order of the items' properties should be also changed
+        softly.assertThat(result.getIt()).containsExactly(goods.toString());
+        softly.assertThat(result.getId()).containsExactly("MATE-1");
+        softly.assertThat(result.getPr()).containsExactly(1190L);
+
```
Comment (heshamMassoud) [inline service/src/test/java/com/commercetools/pspadapter/payone/mapping/CreditCardRequestFactoryTest.java]: why the empty lines?

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java (hunk 6 above)
Comment (ahalberkamp) [reply service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java]: ef466b188

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java (hunk 7 above)
Comment (ahalberkamp) [reply service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java]: 3367ea973

File: service/src/test/java/com/commercetools/pspadapter/payone/mapping/MappingUtilTest.java (hunk 9 above)
Comment (ahalberkamp) [reply service/src/test/java/com/commercetools/pspadapter/payone/mapping/MappingUtilTest.java]: cd344c9ad

File: service/src/test/java/com/commercetools/pspadapter/payone/mapping/CreditCardRequestFactoryTest.java (hunk 10 above)
Comment (ahalberkamp) [reply service/src/test/java/com/commercetools/pspadapter/payone/mapping/CreditCardRequestFactoryTest.java]: 501847cde 2

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java (hunk 8 above)
Comment (ahalberkamp) [reply service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ExtendedAuthorizationRequest.java]: cd344c9

**Pull request merged**